be/src/io/s3_output_stream.cpp

```diff
@@ -0,0 +1,151 @@
+        : _client(std::move(client)),
+          _bucket(std::move(bucket)),
+          _object(std::move(object)),
+          _max_single_part_size(max_single_part_size),
+          _min_upload_part_size(min_upload_part_size),
+          _buffer(),
+          _upload_id(),
+          _etags() {
+    CHECK(_client != nullptr);
+}
+
+Status S3OutputStream::write(const void* data, int64_t size) {
+    _buffer.append(static_cast<const char*>(data), size);
+    if (_upload_id.empty() && _buffer.size() > _max_single_part_size) {
+        RETURN_IF_ERROR(create_multipart_upload());
+        DCHECK(!_upload_id.empty());
+    }
+    if (!_upload_id.empty() && _buffer.size() >= _min_upload_part_size) {
+        RETURN_IF_ERROR(multipart_upload());
+        _buffer.clear();
+    }
+    return Status::OK();
+}
+
+Status S3OutputStream::write_aliased(const void* data, int64_t size) {
+    return Status::NotSupported("S3OutputStream::write_aliased");
+}
+
+Status S3OutputStream::skip(int64_t count) {
+    _buffer.resize(_buffer.size() + count);
+    // Calling `write()` will trigger the uploading checks.
+    return S3OutputStream::write("", 0);
+}
+
+StatusOr<OutputStream::Buffer> S3OutputStream::get_direct_buffer() {
+    return Buffer();
+}
+
+StatusOr<OutputStream::Position> S3OutputStream::get_direct_buffer_and_advance(int64_t size) {
+    auto old_size = _buffer.size();
+    _buffer.resize(old_size + size);
+    return reinterpret_cast<uint8_t*>(_buffer.data()) + old_size;
+}
+
+Status S3OutputStream::close() {
+    if (_upload_id.empty()) {
+        RETURN_IF_ERROR(singlepart_upload());
+    } else {
+        RETURN_IF_ERROR(multipart_upload());
+        RETURN_IF_ERROR(complete_multipart_upload());
+    }
+    _client = nullptr;
+    return Status::OK();
+}
+
+Status S3OutputStream::create_multipart_upload() {
+    Aws::S3::Model::CreateMultipartUploadRequest req;
+    req.SetBucket(_bucket);
+    req.SetKey(_object);
+    Aws::S3::Model::CreateMultipartUploadOutcome outcome = _client->CreateMultipartUpload(req);
+    if (outcome.IsSuccess()) {
+        _upload_id = outcome.GetResult().GetUploadId();
+        return Status::OK();
+    }
+    return Status::IOError(fmt::format("S3: Fail to create multipart upload: {}", outcome.GetError().GetMessage()));
+}
+
+Status S3OutputStream::singlepart_upload() {
+    VLOG(12) << "Uploading s3://" << _bucket << "/" << _object << " vis singlepart upload";
+    DCHECK(_upload_id.empty());
+    if (_buffer.empty()) {
+        return Status::OK();
+    }
+    Aws::S3::Model::PutObjectRequest req;
+    req.SetBucket(_bucket);
+    req.SetKey(_object);
+    req.SetContentLength(static_cast<int64_t>(_buffer.size()));
+    req.SetBody(std::make_shared<Aws::StringStream>(_buffer));
+    Aws::S3::Model::PutObjectOutcome outcome = _client->PutObject(req);
+    if (!outcome.IsSuccess()) {
+        return Status::IOError(fmt::format("S3: Fail to put object: {}", outcome.GetError().GetMessage()));
```

Comment:
why not log the object name at the same time?

Reply:
Updated, PTAL

Reply:
there are 4 `Status::IOError`, only 2 are changed. also is it possible to provide something like 'stream->get_name' to help logging?

Reply:
> 

{Input, Output}Stream isn't always associated with a path or URL, e.g, `ArrayInputStream` and `ArrayOutputStream`. If the name matters,  you can print the names from the outside of the stream, like this:
```
  std::string path = "/path/of/file";
  auto stream = create_output_stream(path);
  auto st = stream.write("content", 7);
  LOG_IF(ERROR, !st.ok()) << "Failt to write " << path << ": " << st;   // Here we print the path associated with the stream
```

Reply:
> > 
> 
> {Input, Output}Streams are not always associated with a path or URL, e.g, `ArrayInputStream` and `ArrayOutputStream`. If the name/path matters, you can save it somewhere else and print it on need.

For example, if we have a CSVScanner that will read data from a file stream, and we want to print the
file patch on failure, we can pass the file path along with the file stream to the CSVScanner:
```
class CSVScanner {
public:
    CSVScanner(std::string file_path, std::shared_ptr<InputStream> file_stream)
        : _file_path(std::move(file_path)), _file_stream(std::move(file_stream)) {}

private:
    std::string _file_path;
    std::shared_ptr<InputStream> _file_stream;
};
```